### PR TITLE
Add spot light with example

### PR DESCRIPTION
**Pull request by DunaevaOlga**
### **What's new:**
A **SpotLight** class has been added that implements conical light in the **direction** and with focusing **p**.

### **Changes and integration:**
To share several different light sources, a virtual function _intensity_ has been added, which calculates the intensity of a given light source at a point. The function is called in _simple.rt_ in the __shadow_ function. By default, the function returns 1, for SpotLight, the value is calculated depending on the distance from the point to the light source.

### CODE REVIEW

File: 15_spotlight.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Example 15: small scene with spot light
```
Comment (daniilsavchuk) [inline 15_spotlight.py]: This file should be in example folder

File: light.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+"""
```
Comment (daniilsavchuk) [inline light.py]: You should apply your changes at original pyrt/light/light.py file

File: simplert.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
+"""
```
Comment (daniilsavchuk) [inline simplert.py]: Changes should be at pyrt/renderer/simplert.py

File: spotlight.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+"""
```
Comment (daniilsavchuk) [inline spotlight.py]: The file should be in a light directory

File: vec3.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
+"""
```
Comment (daniilsavchuk) [inline vec3.py]: Changes should be at math directory

**Review by daniilsavchuk — CHANGES_REQUESTED** [review]
Reorder the structure please

File: pyrt/renderer/simplert.py [hunk 6]
```diff
@@ -50,17 +50,30 @@ def _shade(self, scene: Scene, ray: Ray, hitrecord: HitRecord) -> tuple:
...
             shadowray = Ray(hitrecord.point, light.position - hitrecord.point)
+            if light.name == "SpotLight":
+                radians = np.arccos(light.direction.dot(-shadowray.direction) / (light.direction.length() * shadowray.direction.length()))
+                if(np.degrees(radians) > light.angle):
+                    continue
+                else:
+                    fs += (1 - np.degrees(radians) / (1.5 * light.angle)) / len(scene.lights)
+            else:
+                fs += 1 / len(scene.lights)
+
```
Comment (daniilsavchuk) [inline pyrt/renderer/simplert.py]: Should be a virtual method of SpotLight that doing nothing for a usual light but for the spot - implementing these changes.

File: pyrt/renderer/simplert.py [hunk 7]
```diff
@@ -50,17 +50,30 @@ def _shade(self, scene: Scene, ray: Ray, hitrecord: HitRecord) -> tuple:
...
             local_num_shadow_rays += 1
             for testelement in scene.nodes:
                 if testelement != hitrecord.obj:  # avoid self-intersection
                     if testelement.hitShadow(shadowray):
                         numshadow += 1
                         break
+
+
+        fs *= 0.6 * (len(scene.lights) + 1)
+        fs = np.clip(fs, 0, 1)
```
Comment (daniilsavchuk) [inline pyrt/renderer/simplert.py]: Have no idea what magic at the abstract level class

**Review by daniilsavchuk — CHANGES_REQUESTED** [review]
Please do the following:
1. Fix the comments near the code 
2. Provide wide and MD formatted description of your pull-request. At least it should contain the following topics:
a. Overview
b. Aim 
c. Integration
d. Changes

File: pyrt/light/light.py [hunk 8]
```diff
@@ -13,11 +13,20 @@ class Light(object):
...
-    def __init__(self, name="UNKNOWN LIGHT"):
+    def __init__(self, coef = 1.0, name="UNKNOWN LIGHT"):
         """
         Base Constructor
 
         :param name: Name of Light type, for example "PointLight".
         """
         self.name = name
         self.position = Vec3(0.,0.,0.)
+        self.coef = coef
```
Comment (daniilsavchuk) [inline pyrt/light/light.py]: Undocumented and **coef** - is a strange name for production code(need to name according to a reference or context it is used for)

File: pyrt/light/pointlight.py [hunk 9]
```diff
@@ -12,15 +12,16 @@ class PointLight(Light):
...
-    def __init__(self, position: Vec3) -> None:
+    def __init__(self, position: Vec3, coef: float) -> None:
         """
         Constructor
 
         Not yet complete - params will change in near future!
         :param position: position of the light
         """
-        Light.__init__(self, "PointLight")
+        Light.__init__(self, coef, "PointLight")
```
Comment (daniilsavchuk) [inline pyrt/light/pointlight.py]: You should not touch old one code if it is really not necessary to. So please delete the **coef** param from point-light

File: pyrt/renderer/simplert.py [hunk 10]
```diff
@@ -50,17 +50,21 @@ def _shade(self, scene: Scene, ray: Ray, hitrecord: HitRecord) -> tuple:
 
     def _shadow(self, scene: Scene, hitrecord: HitRecord) -> tuple:
         # is hitpoint in shadow ?
-        fs = 1.0
+        fs = 0.0
```
Comment (daniilsavchuk) [inline pyrt/renderer/simplert.py:53]: What is fs? Why is it zero and why is it divided by 4 after cycle

**Comment by daniilsavchuk** [comment]
image
You still did not fill the description!

**Review by daniilsavchuk — CHANGES_REQUESTED** [review]
See all my comments please

**Comment by daniilsavchuk** [comment]
By the task provided, we are going to implement the spotlight, not conic. Conic light is a specific case of the spotlight. Please improve your model at least using this method: https://books.google.ru/books?id=rHuc4yQ7TDoC&pg=PA204&lpg=PA204&dq=spotlight+model+graphics&source=bl&ots=hvCGLmxlB2&sig=ACfU3U0fXX_v4dXhJibG2AycEE4nl1ZjrQ&hl=ru&sa=X&ved=2ahUKEwjNtc3lwJ3oAhVG06YKHVbvBJcQ6AEwCnoECAoQAQ#v=onepage&q=spotlight%20model%20graphics&f=false

Perfectly if you would like to see the most realistic picture - use the Fresnel distortions too. 

Thank you in advance. Anyways looks like a great job!

File: pyrt/renderer/simplert.py (hunk 10 above)
Comment (DunaevaOlga) [reply pyrt/renderer/simplert.py:53]: fs - total intensity at a point
In the absence of light sources is equal to 0
If the point is in the shadow of the object, then the intensity drops by 4 times (it was in the project)

File: pyrt/light/pointlight.py (hunk 9 above)
Comment (DunaevaOlga) [reply pyrt/light/pointlight.py]: This is the intensity of the light source as a parameter, without this coefficient the picture is not so beautiful, but now it is deleted everywhere

File: pyrt/light/spotlight.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+    def intensity(self, shadowray):
+        """
+        Point intensity calculation:
+        param shadowray: ray from light to hitrecord point
+        """
+
+        c1 = 0.0
+        c2 = 0.1
+        c3 = 0.02
```
Comment (daniilsavchuk) [inline pyrt/light/spotlight.py]: Make them configurable from spotlight constructor with default values

**Review by daniilsavchuk — CHANGES_REQUESTED** [review]
1. Improve PR message
2. Correct last comment 
3. Create PR to the source repo

**Pull request merged**